### PR TITLE
Add replacement proposal leader

### PR DESCRIPTION
**Pull request by peterwillcn**
Signed-off-by: Xiaobo Tian <peterwillcn@gmail.com>

### CODE REVIEW

File: blockchain/txvalidator.go [hunk 1]
```diff
@@ -2160,6 +2160,42 @@ func (b *BlockChain) checkCRCProposalTransaction(txn *Transaction,
 		return errors.New("budgets exceeded the maximum limit")
 	}
 
+	if proposal.ProposalType == payload.ChangeProposalOwner {
+		if !b.crCommittee.IsInElectionPeriod() || b.crCommittee.IsInVotingPeriod(blockHeight) {
```
Comment (xiaomingfuckeasylife) [inline blockchain/txvalidator.go:2164]: change proposal owner transaction should be in election period and not in voting period . should it be like `b.crCommittee.IsInElectionPeriod() && !b.crCommittee.IsInVotingPeriod(blockHeight)`

File: blockchain/txvalidator.go [hunk 2]
```diff
@@ -2160,6 +2160,42 @@ func (b *BlockChain) checkCRCProposalTransaction(txn *Transaction,
...
+			return errors.New("cr proposal tx must be at election Period and not during voting period")
+		}
+
+		proposalState := b.crCommittee.GetProposal(proposal.PreviousHash)
+		if proposalState.Status != crstate.VoterAgreed {
+			return errors.New("proposal status is not VoterAgreed")
+		}
+
+		if b.crCommittee.ExistProposal(proposal.Hash()) {
```
Comment (xiaomingfuckeasylife) [inline blockchain/txvalidator.go:2173]: current proposal is not exist at the moment , we don't need this if statement i think.

File: blockchain/txvalidator.go [hunk 3]
```diff
@@ -2160,6 +2160,42 @@ func (b *BlockChain) checkCRCProposalTransaction(txn *Transaction,
...
+			return errors.New("invalid owner")
+		}
+
+		if newPublicKey == publicKey {
+			return errors.New("cr new did must be different from the previous one")
+		}
+
+		newCode, err := contract.CreateStandardRedeemScript(newPublicKey)
+		did, _ := getDIDByCode(newCode)
+		if crMember := b.crCommittee.GetMember(*did); crMember == nil {
```
Comment (xiaomingfuckeasylife) [inline blockchain/txvalidator.go:2193]: cr member also must be MemberElected

File: cr/state/proposalmanager.go [hunk 4]
```diff
@@ -193,6 +193,14 @@ func (p *ProposalManager) updateProposals(height uint32,
 					unusedAmount += getProposalTotalBudgetAmount(v.Proposal)
 					continue
 				}
+				if v.Proposal.ProposalType == payload.ChangeProposalOwner {
+					proposal := p.getProposal(v.Proposal.PreviousHash)
+					p.history.Append(height, func() {
+						proposal.ProposalOwner = v.Proposal.NewOwnerPublicKey
```
Comment (xiaomingfuckeasylife) [inline cr/state/proposalmanager.go:199]: and the recipient also need to be updated if it is not blank

File: blockchain/txvalidator.go (hunk 1 above)
Comment (peterwillcn) [reply blockchain/txvalidator.go:2164]: return error ?

File: cr/state/proposalmanager.go [hunk 5]
```diff
@@ -193,6 +193,20 @@ func (p *ProposalManager) updateProposals(height uint32,
...
 					continue
 				}
+				if v.Proposal.ProposalType == payload.ChangeProposalOwner {
+					proposal := p.getProposal(v.Proposal.PreviousHash)
+					originRecipient := proposal.Proposal.Recipient
+					emptyUint168 := common.Uint168{}
+					p.history.Append(height, func() {
+						proposal.ProposalOwner = v.Proposal.NewOwnerPublicKey
+						if v.Proposal.Recipient != emptyUint168 {
+							proposal.Proposal.Recipient = v.Proposal.Recipient
```
Comment (RainFallsSilent) [inline cr/state/proposalmanager.go:203]: need to reset recipient as ProposalOwner
1.add 'Recpient' field into ProposalState
2.set value of ProposalState Recipient when registering proposal
3.reset value of ProposalState Recipient when changing proposal owner
4.adjust all places where this field is used in check transactions functions，modify from 'proposalState.proposal.Recipient' to 'proposalState.Recipient'

File: blockchain/txvalidator.go [hunk 6]
```diff
@@ -2160,6 +2160,38 @@ func (b *BlockChain) checkCRCProposalTransaction(txn *Transaction,
 		return errors.New("budgets exceeded the maximum limit")
 	}
 
+	if proposal.ProposalType == payload.ChangeProposalOwner {
+		proposalState := b.crCommittee.GetProposal(proposal.PreviousHash)
+		if proposalState.Status != crstate.VoterAgreed {
```
Comment (Houshoupei84) [inline blockchain/txvalidator.go:2165]: proposalState might be nil

File: blockchain/txvalidator.go [hunk 7]
```diff
@@ -2160,6 +2160,38 @@ func (b *BlockChain) checkCRCProposalTransaction(txn *Transaction,
...
+		if err != nil {
+			return errors.New("invalid owner")
+		}
+
+		if newPublicKey == publicKey {
+			return errors.New("cr new did must be different from the previous one")
+		}
+
+		newCode, err := contract.CreateStandardRedeemScript(newPublicKey)
+		did, _ := getDIDByCode(newCode)
```
Comment (Houshoupei84) [inline blockchain/txvalidator.go:2184]: The error on the main chain should not be omitted

**Pull request merged**